### PR TITLE
Show value of locals in debugger tooltip (#1737)

### PR DESCRIPTION
**Pull request by mallt**
Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Added the value of the local to the mouse tooltip in the debugger.

### CODE REVIEW

File: cider-mode.el [hunk 1]
```diff
@@ -671,7 +671,11 @@ See \(info \"(elisp) Special Properties\")"
             (goto-char pos)
             (when-let ((sym (cider-symbol-at-point)))
               (if (member sym (get-text-property (point) 'cider-locals))
-                  (format "`%s' is a local" sym)
+                  (concat (format "`%s' is a local" sym)
+                          (nrepl-dbind-response cider--debug-mode-response (debug-value input-type locals)
+                            (let* ((filtered-locals (delq nil (mapcar (lambda (l) (member sym l)) locals)))
+                                   (local-val (nth 1 (car filtered-locals))))
```
Comment (Malabarba) [inline cider-mode.el]: IIUC, these two lines are sort of reimplementing `nrepl-dict-get`.

File: cider-mode.el [hunk 2]
```diff
@@ -671,7 +671,11 @@ See \(info \"(elisp) Special Properties\")"
             (goto-char pos)
             (when-let ((sym (cider-symbol-at-point)))
               (if (member sym (get-text-property (point) 'cider-locals))
-                  (format "`%s' is a local" sym)
+                  (concat (format "`%s' is a local" sym)
+                          (nrepl-dbind-response cider--debug-mode-response (debug-value input-type locals)
```
Comment (Malabarba) [inline cider-mode.el]: wrap this in a `(when cider--debug-mode ...)`

File: cider-mode.el [hunk 3]
```diff
@@ -671,7 +671,11 @@ See \(info \"(elisp) Special Properties\")"
...
+                            (let* ((filtered-locals (delq nil (mapcar (lambda (l) (member sym l)) locals)))
+                                   (local-val (nth 1 (car filtered-locals))))
+                              (format " (value: %s)" local-val))))
```
Comment (Malabarba) [inline cider-mode.el]: Replace this format string with `" with value:\n%s"`.

**Comment by Malabarba** [comment]
Other than my 3 comments, the code looks good to me.
Change your commit message to `[Fix #1737] Show value of locals in debugger tooltip` and add a changelog entry (under `### Changes`).

**Comment by mallt** [comment]
Thank you for the comments!
What is the best way to fix the "reference to free variable `cider--debug-mode-response'" error? Should I require 'cider-debug?

File: cider-mode.el [hunk 4]
```diff
@@ -671,7 +674,12 @@ See \(info \"(elisp) Special Properties\")"
             (goto-char pos)
             (when-let ((sym (cider-symbol-at-point)))
               (if (member sym (get-text-property (point) 'cider-locals))
-                  (format "`%s' is a local" sym)
+                  (concat (format "`%s' is a local" sym)
+                          (when cider--debug-mode
+                            (let* ((locals (nrepl-dict-get cider--debug-mode-response "locals"))
+                                   (filtered-locals (delq nil (mapcar (lambda (l) (member sym l)) locals)))
```
Comment (bbatsov) [inline cider-mode.el]: Can't you use `seq-first` or something like this?

File: CHANGELOG.md [hunk 5]
```diff
@@ -26,6 +26,7 @@
 * Extremely long overlays are truncated and `cider-inspect-last-result` is recommended.
 * Signal `user-error` instead of `error` on jack-in if a project type is not supported.
 * Users with `boot.sh` instead of `boot` should customize `cider-boot-command` instead of relying on automatic detection.
+* [#1737](https://github.com/clojure-emacs/cider/issues/1737): Show value of locals in debugger tooltip
```
Comment (bbatsov) [inline CHANGELOG.md]: This should end with a `.`.

File: cider-mode.el (hunk 4 above)
Comment (Malabarba) [inline cider-mode.el]: Ah, took me a while, but now I understand what this is trying to do.
What you actually want to use here is `assoc`.

    (local-val (cadr (assoc sym locals)))

**Comment by mallt** [comment]
Thanks for the help! Please let me know if this version is ok to merge.

**Comment by Malabarba** [comment]
Replace `(local-val (cadr (seq-find (lambda (l) (member sym l)) locals)))`
with `(local-val (cadr (assoc sym locals)))`
and we're good to merge! :+1:

**Comment by Malabarba** [comment]
Awesome. Thanks for the help!

**Pull request merged**